include/seqan3/alignment/pairwise/alignment_algorithm.hpp

```diff
@@ -387,38 +363,66 @@ class alignment_algorithm :
                         second_range_t & second_range,
                         cache_t & cache)
     {
-        using std::get;
+        using zip_column_t = decltype(views::zip(*this->score_matrix_iter, *this->trace_matrix_iter));
+        using column_iter_t = std::ranges::iterator_t<decltype(std::declval<zip_column_t>() | views::take_exactly(1))>;
+
+        if (is_fst_range_empty)  // Already tracked the max in initialisation.
+            return;
+
         auto const & score_scheme = get<align_cfg::scoring>(*cfg_ptr).value;
-        std::ranges::for_each(first_range, [&, this](auto seq1_value)
-        {
-            // Move internal matrix to next column.
-            this->go_next_column();
 
-            auto col = this->current_column();
-            this->init_row_cell(*std::ranges::begin(col), cache);
+        column_iter_t column_iter{};
+        ranges::for_each(first_range, [&, this] (auto const seq1_value)
```

Comment:
std::for_each requires common range, which is not required. I could do everything with a normal for loop, but does it make it more clear? It certainly does not reduce code. 🤷‍♂

Reply:
from [fluent C++](https://www.fluentcpp.com/2018/03/30/is-stdfor_each-obsolete/)

> for_each is a tool for raising the level of abstraction of a range-based for loop. And showing the inside of the lambda within the call to for_each kills this purpose.

So if you save the lambda somewhere because you want to reuse it, ir its a aisngle line of code, I am in favour of std::ranges::for_each.
But if you just put the lambda inside the for_each body I think a range based for loop is more clear.

(Also I am not sure if introducing a lambda that captures many variables by reference may have any overhead?)

Reply:
> So if you save the lambda somewhere because you want to reuse it, ir its a aisngle line of code, I am in favour of std::ranges::for_each.
But if you just put the lambda inside the for_each body I think a range based for loop is more clear.

I wouldn’t say that in general. At least the code above is expressive since it does tell you in the first line what are the variables and in the following body how it is used.. But my concern is that there are no 0-cost abstractions. So that could be a reason to make it a regular for loop. Maybe that is also expresses a clearer message, of hey, this is a hot kernel. Don’t mess with it. It is done on purpose. On the other hand there might be different kernels called depending on vectorisation or not. We’ll see 👨‍💻

Reply:
I don't understand exactly which of your answers supports or is against my proposal.

> Maybe that is also expresses a clearer message, of hey, this is a hot kernel

which one? either way I disagree because the lambda is defined within the body of std::for_each so **it is** a range based for loop with an extra capture.

> On the other hand there might be different kernels called depending on vectorisation or not

Well then maybe the hot kernel should be outsourced into it's own function and **then** you can call std::ranges::for_each with this function.

Reply:
> which one? either way I disagree because the lambda is defined within the body of std::for_each so it is a range based for loop with an extra capture. 
Well then maybe the hot kernel should be outsourced into it's own function and then you can call std::ranges::for_each with this function.

The point is using std::for_each adds a level of abstraction. That will imply costs. Thus, the only argument against using it would be to use less abstractions as possible. Outsourcing the lambda makes only sense if it is used multiple times or if there are conditional differences calling the same for_each with different functions. None of it applies to the current PR. I am reluctant to make structural changes just because of some personal preferences. DOing some micro optimisation doesn't make a lot of sense since we do not know how the code looks like once the vectorisation is incorporated. If the final design is known, we can start optimising kernels by switching between regular for loops and for_each algorithms. Until then it is a waste of time to do so.

Reply:
Ok, after much debating and clearly missing all the points. The short story is I am changing this here to range based for loops. Makes total sense in this case anyway